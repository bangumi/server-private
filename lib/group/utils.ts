import { db, op, schema } from '@app/drizzle';

/** Is user member of group */
export async function isMemberInGroup(userID: number, groupID: number): Promise<boolean> {
  const [d] = await db
    .select({ uid: schema.chiiGroupMembers.uid, gid: schema.chiiGroupMembers.gid })
    .from(schema.chiiGroupMembers)
    .where(
      op.and(
        op.eq(schema.chiiGroupMembers.uid, userID),
        op.eq(schema.chiiGroupMembers.gid, groupID),
      ),
    )
    .limit(1);

  return Boolean(d);
}
