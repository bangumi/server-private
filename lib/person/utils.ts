import { db, op, schema } from '@app/drizzle';

import type { PersonCat } from './type';

export async function getPersonCollect(
  cat: PersonCat,
  userID: number,
  personID: number,
): Promise<number | null> {
  const [d] = await db
    .select()
    .from(schema.chiiPersonCollects)
    .where(
      op.and(
        op.eq(schema.chiiPersonCollects.cat, cat),
        op.eq(schema.chiiPersonCollects.uid, userID),
        op.eq(schema.chiiPersonCollects.mid, personID),
      ),
    )
    .limit(1);
  if (!d) {
    return null;
  }
  return d.createdAt;
}
