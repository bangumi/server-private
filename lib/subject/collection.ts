export enum CollectionType {
  Wish = 1,
  Collect = 2,
  Doing = 3,
  OnHold = 4,
  Dropped = 5,
}

export const CollectionTypeValues = new Set([1, 2, 3, 4, 5]);
export const CollectionTypeProfileValues = new Set([1, 2]);
