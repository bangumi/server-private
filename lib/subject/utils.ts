import * as php from '@trim21/php-serialize';
import { DateTime } from 'luxon';

import { db, decr, incr, op, type orm, schema, type Txn } from '@app/drizzle';

import { type CollectionType, SubjectType, type UserEpisodeStatusItem } from './type';
import { EpisodeCollectionStatus, getCollectionTypeField } from './type';

/** 更新条目收藏计数，需要在事务中执行 */
export async function updateSubjectCollection(
  t: Txn,
  subjectID: number,
  newType: CollectionType,
  oldType?: CollectionType,
) {
  if (oldType && oldType === newType) {
    return;
  }
  const toUpdate: Record<string, op.SQL> = {};
  toUpdate[getCollectionTypeField(newType)] = incr(
    schema.chiiSubjects[getCollectionTypeField(newType)],
  );
  if (oldType) {
    toUpdate[getCollectionTypeField(oldType)] = decr(
      schema.chiiSubjects[getCollectionTypeField(oldType)],
    );
  }
  await t
    .update(schema.chiiSubjects)
    .set(toUpdate)
    .where(op.eq(schema.chiiSubjects.id, subjectID))
    .limit(1);
}

function getRatingField(rate: number) {
  return [
    undefined,
    'rate1',
    'rate2',
    'rate3',
    'rate4',
    'rate5',
    'rate6',
    'rate7',
    'rate8',
    'rate9',
    'rate10',
  ][rate];
}

/** 更新条目评分，需要在事务中执行 */
export async function updateSubjectRating(
  t: Txn,
  subjectID: number,
  oldRate: number,
  newRate: number,
) {
  if (oldRate === newRate) {
    return;
  }
  const newField = getRatingField(newRate);
  const oldField = getRatingField(oldRate);
  const toUpdate: Record<string, op.SQL> = {};
  if (newField) {
    const field = newField as keyof orm.ISubjectFields;
    toUpdate[field] = incr(schema.chiiSubjectFields[field]);
  }
  if (oldField) {
    const field = oldField as keyof orm.ISubjectFields;
    toUpdate[field] = decr(schema.chiiSubjectFields[field]);
  }
  if (Object.keys(toUpdate).length === 0) {
    return;
  }
  await t
    .update(schema.chiiSubjectFields)
    .set(toUpdate)
    .where(op.eq(schema.chiiSubjectFields.id, subjectID))
    .limit(1);
}

/** 更新条目剧集进度，需要在事务中执行 */
export async function updateSubjectEpisodeProgress(
  t: Txn,
  userID: number,
  subjectID: number,
  episodeID: number,
  type: number,
): Promise<number> {
  const [current] = await t
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    );
  let watchedEpisodes = 0;
  if (current?.status) {
    const epStatusList = parseSubjectEpStatus(current.status);
    epStatusList[episodeID] = {
      eid: episodeID,
      type,
    };
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t
      .update(schema.chiiEpStatus)
      .set({ status: newStatus, updatedAt: DateTime.now().toUnixInteger() })
      .where(op.eq(schema.chiiEpStatus.id, current.id))
      .limit(1);
  } else {
    const epStatusList: Record<number, UserEpisodeStatusItem> = {};
    epStatusList[episodeID] = {
      eid: episodeID,
      type,
    };
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t.insert(schema.chiiEpStatus).values({
      uid: userID,
      sid: subjectID,
      status: newStatus,
      updatedAt: DateTime.now().toUnixInteger(),
    });
  }
  return watchedEpisodes;
}

/** 标记条目剧集为已观看，需要在事务中执行 */
export async function markEpisodesAsWatched(
  t: Txn,
  userID: number,
  subjectID: number,
  episodeIDs: number[],
  revertOthers = false,
): Promise<number> {
  const epStatusList: Record<number, UserEpisodeStatusItem> = {};
  for (const episodeID of episodeIDs) {
    epStatusList[episodeID] = {
      eid: episodeID,
      type: EpisodeCollectionStatus.Done,
    };
  }
  const [current] = await t
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    );
  let watchedEpisodes = 0;
  if (current?.status) {
    const oldList = parseSubjectEpStatus(current.status);
    for (const x of Object.values(oldList)) {
      if (episodeIDs.includes(x.eid)) {
        continue;
      }
      if (revertOthers && x.type === EpisodeCollectionStatus.Done) {
        epStatusList[x.eid] = {
          eid: x.eid,
          type: EpisodeCollectionStatus.None,
        };
      } else {
        epStatusList[x.eid] = x;
      }
    }
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t
      .update(schema.chiiEpStatus)
      .set({ status: newStatus, updatedAt: DateTime.now().toUnixInteger() })
      .where(op.eq(schema.chiiEpStatus.id, current.id))
      .limit(1);
  } else {
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t.insert(schema.chiiEpStatus).values({
      uid: userID,
      sid: subjectID,
      status: newStatus,
      updatedAt: DateTime.now().toUnixInteger(),
    });
  }
  return watchedEpisodes;
}

export function parseSubjectEpStatus(status: string): Record<number, UserEpisodeStatusItem> {
  const result: Record<number, UserEpisodeStatusItem> = {};
  if (!status) {
    return result;
  }
  const epStatusList = php.parse(status) as Record<number, UserEpisodeStatusItem>;
  for (const x of Object.values(epStatusList)) {
    result[x.eid] = x;
  }
  return result;
}

export async function getEpStatus(
  userID: number,
  subjectID: number,
): Promise<Record<number, UserEpisodeStatusItem>> {
  const [data] = await db
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    )
    .limit(1);
  if (!data) {
    return {};
  }
  return parseSubjectEpStatus(data.status);
}

/** 完成条目进度，需要在事务中执行 */
export async function completeSubjectProgress(
  t: Txn,
  userID: number,
  subject: orm.ISubject,
  interest: Partial<orm.ISubjectInterest>,
) {
  if (subject.eps > 0) {
    interest.epStatus = subject.eps;
  }
  if (subject.volumes > 0) {
    interest.volStatus = subject.volumes;
  }
  if ([SubjectType.Anime, SubjectType.Real].includes(subject.typeID)) {
    const episodes = await t
      .select({ id: schema.chiiEpisodes.id })
      .from(schema.chiiEpisodes)
      .where(
        op.and(
          op.eq(schema.chiiEpisodes.subjectID, subject.id),
          op.eq(schema.chiiEpisodes.type, 0),
          op.eq(schema.chiiEpisodes.ban, 0),
        ),
      );
    const episodeIDs = episodes.map((e) => e.id);
    if (episodeIDs.length > 0) {
      await markEpisodesAsWatched(t, userID, subject.id, episodeIDs);
    }
  }
}
