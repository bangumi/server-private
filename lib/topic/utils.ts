import type * as orm from '@app/drizzle/orm.ts';
import { HotGroups } from '@app/lib/group/consts.ts';

export function scoredUpdateTime(timestamp: number, topic: orm.IGroupTopic): number {
  if (HotGroups.includes(topic.gid) && topic.replies > 0) {
    const createdAt = topic.createdAt;
    const createdHours = (timestamp - createdAt) / 3600;
    const gravity = 1.8;
    const baseScore = (Math.pow(createdHours + 0.1, gravity) / topic.replies) * 200;
    const scoredLastPost = Math.trunc(timestamp - baseScore);
    return Math.min(scoredLastPost, timestamp);
  }

  return timestamp;
}
