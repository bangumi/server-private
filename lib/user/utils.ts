import { db, op, schema } from '@app/drizzle';
import { UnexpectedNotFoundError } from '@app/lib/error.ts';
import redis from '@app/lib/redis.ts';
import * as fetcher from '@app/lib/types/fetcher.ts';
import type * as res from '@app/lib/types/res.ts';
import {
  getFollowersCacheKey,
  getFriendsCacheKey,
  getRelationCacheKey,
} from '@app/lib/user/cache.ts';

/** Cached: Get friend ids of user(uid) */
export async function fetchFriends(uid?: number): Promise<number[]> {
  if (!uid) {
    return [];
  }

  const cached = await redis.get(getFriendsCacheKey(uid));
  if (cached) {
    return JSON.parse(cached) as number[];
  }

  const friends = await db
    .select({ fid: schema.chiiFriends.fid })
    .from(schema.chiiFriends)
    .where(op.eq(schema.chiiFriends.uid, uid));
  const result = friends.map((x) => x.fid);
  await redis.setex(getFriendsCacheKey(uid), 3600, JSON.stringify(result));
  return result;
}

/** Cached: Get follower ids of user(uid) */
export async function fetchFollowers(uid?: number): Promise<number[]> {
  if (!uid) {
    return [];
  }

  const cached = await redis.get(getFollowersCacheKey(uid));
  if (cached) {
    return JSON.parse(cached) as number[];
  }

  const followers = await db
    .select({ uid: schema.chiiFriends.uid })
    .from(schema.chiiFriends)
    .where(op.eq(schema.chiiFriends.fid, uid));
  const result = followers.map((x) => x.uid);
  await redis.setex(getFollowersCacheKey(uid), 3600, JSON.stringify(result));
  return result;
}

/** Cached: Is user(another) is friend of user(uid) */
export async function isFriends(uid: number, another: number): Promise<boolean> {
  const cached = await redis.get(getRelationCacheKey(uid, another));
  if (cached) {
    return cached === '1';
  }

  const [d] = await db
    .select({ uid: schema.chiiFriends.uid, fid: schema.chiiFriends.fid })
    .from(schema.chiiFriends)
    .where(op.and(op.eq(schema.chiiFriends.uid, uid), op.eq(schema.chiiFriends.fid, another)))
    .limit(1);
  const result = d ? 1 : 0;
  await redis.setex(getRelationCacheKey(uid, another), 3600, result);
  return result === 1;
}

export function ghostUser(uid: number): res.ISlimUser {
  return {
    id: 0,
    username: uid.toString(),
    nickname: `deleted or missing user ${uid}`,
    avatar: {
      small: '',
      medium: '',
      large: '',
    },
    group: 0,
    sign: '',
    joinedAt: 0,
  };
}

/** Cached */
export async function fetchUserX(uid: number): Promise<res.ISlimUser> {
  const user = await fetcher.fetchSlimUserByID(uid);
  if (!user) {
    throw new UnexpectedNotFoundError(`user ${uid} not found`);
  }
  return user;
}
