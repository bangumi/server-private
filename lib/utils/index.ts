import * as crypto from 'node:crypto';

import { customAlphabet } from 'nanoid/async';

const base62Chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
if (base62Chars.length !== 62) {
  throw new TypeError('characters is not 62 length');
}

const generator = customAlphabet(base62Chars, 32);

export const randomBase62String = (size: number) => generator(size);

export async function randomBytes(size: number): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    crypto.randomBytes(size, (err, buf) => {
      if (err) {
        reject(err);
      } else {
        resolve(buf);
      }
    });
  });
}

/**
 * Parse string as int, strictly
 *
 * 避免出现 `parseInt('1abc') === 1` 的问题
 */
export function intval(value: string | number): number {
  const v = parseIntStrict(value);
  if (v === undefined) {
    throw new Error(`"${value}" is not a valid integer`);
  }
  return v;
}

function parseIntStrict(integer: string | number) {
  if (typeof integer === 'number') {
    return Number.isInteger(integer) ? integer : undefined;
  }

  const n = Number(integer);

  if (Number.isInteger(n)) {
    return n;
  }
}

export function md5(s: string) {
  return crypto.createHash('md5').update(s).digest('hex');
}
