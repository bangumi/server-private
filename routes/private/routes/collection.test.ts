import { describe, expect, test } from 'vitest';

import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './collection.ts';

describe('user collection', () => {
  test('should get summary', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/users/382951/collections/summary',
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subjects', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/users/382951/collections/subjects',
      query: { subjectType: '2', type: '2', limit: '1', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get characters', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/users/1/collections/characters',
      query: { limit: '1', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get persons', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/users/1/collections/persons',
      query: { limit: '1', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get indexes', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/users/1/collections/indexes',
      query: { limit: '1', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get created indexes', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/users/14127/indexes',
      query: { limit: '1', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });
});
