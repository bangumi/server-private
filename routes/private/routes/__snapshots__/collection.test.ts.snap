// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`user collection > should get subjects 1`] = `
Object {
  "data": Array [
    Object {
      "comment": "test comment",
      "epStatus": 23,
      "private": false,
      "rate": 0,
      "subject": Object {
        "airtime": Object {
          "date": "2008-04-06",
          "month": 4,
          "weekday": 7,
          "year": 2008,
        },
        "collection": Object {
          "1": 622,
          "2": 13216,
          "3": 147,
          "4": 224,
          "5": 115,
        },
        "eps": 25,
        "id": 8,
        "images": Object {
          "common": "https://lain.bgm.tv/pic/cover/c/c9/f0/8_wK0z3.jpg",
          "grid": "https://lain.bgm.tv/pic/cover/g/c9/f0/8_wK0z3.jpg",
          "large": "https://lain.bgm.tv/pic/cover/l/c9/f0/8_wK0z3.jpg",
          "medium": "https://lain.bgm.tv/pic/cover/m/c9/f0/8_wK0z3.jpg",
          "small": "https://lain.bgm.tv/pic/cover/s/c9/f0/8_wK0z3.jpg",
        },
        "infobox": Object {
          "Copyright": Array [
            Object {
              "v": "（C）2006 SUNRISE inc./MBS",
            },
          ],
          "中文名": Array [
            Object {
              "v": "Code Geass 反叛的鲁路修R2",
            },
          ],
          "人物原案": Array [
            Object {
              "v": "CLAMP",
            },
          ],
          "人物设定": Array [
            Object {
              "v": "木村貴宏",
            },
          ],
          "其他": Array [
            Object {
              "v": "",
            },
          ],
          "其他电视台": Array [
            Object {
              "v": "",
            },
          ],
          "别名": Array [
            Object {
              "v": "叛逆的鲁路修R2",
            },
            Object {
              "v": "Code Geass: Hangyaku no Lelouch R2",
            },
            Object {
              "v": "叛逆的勒鲁什R2",
            },
            Object {
              "v": "叛逆的鲁鲁修R2",
            },
            Object {
              "v": "コードギアス 反逆のルルーシュR2",
            },
            Object {
              "v": "Code Geass: Lelouch of the Rebellion R2",
            },
            Object {
              "v": "叛逆的勒路什R2",
            },
          ],
          "动画制作": Array [
            Object {
              "v": "サンライズ",
            },
          ],
          "官方网站": Array [
            Object {
              "v": "http://www.geass.jp/r2/",
            },
          ],
          "导演": Array [
            Object {
              "v": "谷口悟朗",
            },
          ],
          "摄影监督": Array [
            Object {
              "v": "大矢創太",
            },
          ],
          "播放电视台": Array [
            Object {
              "v": "每日放送",
            },
          ],
          "播放结束": Array [
            Object {
              "v": "2008年9月28日",
            },
          ],
          "放送开始": Array [
            Object {
              "v": "2008年4月6日",
            },
          ],
          "放送星期": Array [
            Object {
              "v": "",
            },
          ],
          "系列构成": Array [
            Object {
              "v": "大河内一楼",
            },
          ],
          "美术监督": Array [
            Object {
              "v": "菱沼由典",
            },
          ],
          "色彩设计": Array [
            Object {
              "v": "岩沢れい子",
            },
          ],
          "话数": Array [
            Object {
              "v": "25",
            },
          ],
          "音乐": Array [
            Object {
              "v": "中川幸太郎、黒石ひとみ",
            },
          ],
          "音乐制作": Array [
            Object {
              "v": "AUDIO PLANNING U",
            },
          ],
          "音响监督": Array [
            Object {
              "v": "浦上靖夫、井澤基",
            },
          ],
        },
        "locked": false,
        "metaTags": Array [],
        "name": "コードギアス 反逆のルルーシュR2",
        "nameCN": "Code Geass 反叛的鲁路修R2",
        "nsfw": false,
        "platform": Object {
          "alias": "tv",
          "enableHeader": true,
          "id": 1,
          "order": 0,
          "type": "TV",
          "typeCN": "TV",
          "wikiTpl": "TVAnime",
        },
        "rating": Object {
          "count": Array [
            44,
            15,
            32,
            66,
            145,
            457,
            1472,
            3190,
            2640,
            1377,
          ],
          "score": 8.19,
          "total": 9438,
        },
        "redirect": 0,
        "series": false,
        "seriesEntry": 0,
        "summary": "　　“东京决战”一年后，布里塔尼亚少年鲁路修在11区（原日本国）过着平凡的学生生活。但是，鲁路修与弟弟罗洛的一次出行，遇到了黑色骑士团的余党。在与少女C.C再次结成契约之后，尘封的记忆摆在了鲁路修的面前。
　　身为帝国王子的鲁路修，为了建立人人平等的世界、让妹妹娜娜丽幸福的世界，而使用GEASS，令人绝对服从的力量，带领黑色骑士团向帝国宣战。带上假面化名ZERO的他，却在一年前的“东京决战”中被好友朱雀击败，被帝国皇帝抹去了记忆。
　　现在，恢复记忆的鲁路修不仅要面对帝国的强大军事力量，更要面对虚假的弟弟罗洛、失踪的妹妹娜娜丽、不知敌友的中华联盟、内部出现分歧的黑色骑士团。面对内忧外患，鲁路修走上了GEASS之力的诅咒——孤独的王之路。 ",
        "type": 2,
        "volumes": 0,
      },
      "tags": Array [
        "SUNRISE",
      ],
      "type": 2,
      "updatedAt": 1639569371,
      "volStatus": 0,
    },
  ],
  "total": 1,
}
`;

exports[`user collection > should get summary 1`] = `
Object {
  "subject": Object {
    "1": Object {
      "counts": Object {
        "1": 0,
        "2": 0,
        "3": 0,
        "4": 1,
        "5": 0,
      },
      "details": Object {
        "1": Array [],
        "2": Array [],
      },
    },
    "2": Object {
      "counts": Object {
        "1": 0,
        "2": 1,
        "3": 0,
        "4": 0,
        "5": 0,
      },
      "details": Object {
        "1": Array [],
        "2": Array [
          Object {
            "id": 8,
            "images": Object {
              "common": "https://lain.bgm.tv/pic/cover/c/c9/f0/8_wK0z3.jpg",
              "grid": "https://lain.bgm.tv/pic/cover/g/c9/f0/8_wK0z3.jpg",
              "large": "https://lain.bgm.tv/pic/cover/l/c9/f0/8_wK0z3.jpg",
              "medium": "https://lain.bgm.tv/pic/cover/m/c9/f0/8_wK0z3.jpg",
              "small": "https://lain.bgm.tv/pic/cover/s/c9/f0/8_wK0z3.jpg",
            },
            "locked": false,
            "name": "コードギアス 反逆のルルーシュR2",
            "nameCN": "Code Geass 反叛的鲁路修R2",
            "nsfw": false,
            "type": 2,
          },
        ],
      },
    },
    "3": Object {
      "counts": Object {
        "1": 0,
        "2": 0,
        "3": 1,
        "4": 0,
        "5": 0,
      },
      "details": Object {
        "1": Array [],
        "2": Array [],
      },
    },
    "4": Object {
      "counts": Object {
        "1": 21,
        "2": 0,
        "3": 0,
        "4": 0,
        "5": 1,
      },
      "details": Object {
        "1": Array [
          Object {
            "id": 4,
            "images": Object {
              "common": "https://lain.bgm.tv/pic/cover/c/a8/7f/4_cMMK5.jpg",
              "grid": "https://lain.bgm.tv/pic/cover/g/a8/7f/4_cMMK5.jpg",
              "large": "https://lain.bgm.tv/pic/cover/l/a8/7f/4_cMMK5.jpg",
              "medium": "https://lain.bgm.tv/pic/cover/m/a8/7f/4_cMMK5.jpg",
              "small": "https://lain.bgm.tv/pic/cover/s/a8/7f/4_cMMK5.jpg",
            },
            "locked": false,
            "name": "メタルスラッグ7",
            "nameCN": "合金弹头7",
            "nsfw": false,
            "type": 4,
          },
          Object {
            "id": 84399,
            "images": Object {
              "common": "https://lain.bgm.tv/pic/cover/c/85/3f/84399_y0J0e.jpg",
              "grid": "https://lain.bgm.tv/pic/cover/g/85/3f/84399_y0J0e.jpg",
              "large": "https://lain.bgm.tv/pic/cover/l/85/3f/84399_y0J0e.jpg",
              "medium": "https://lain.bgm.tv/pic/cover/m/85/3f/84399_y0J0e.jpg",
              "small": "https://lain.bgm.tv/pic/cover/s/85/3f/84399_y0J0e.jpg",
            },
            "locked": false,
            "name": "Criminal Minds (Season 9)",
            "nameCN": "犯罪心理 第九季",
            "nsfw": false,
            "type": 6,
          },
          Object {
            "id": 104833,
            "images": Object {
              "common": "https://lain.bgm.tv/pic/cover/c/35/eb/104833_CAd6k.jpg",
              "grid": "https://lain.bgm.tv/pic/cover/g/35/eb/104833_CAd6k.jpg",
              "large": "https://lain.bgm.tv/pic/cover/l/35/eb/104833_CAd6k.jpg",
              "medium": "https://lain.bgm.tv/pic/cover/m/35/eb/104833_CAd6k.jpg",
              "small": "https://lain.bgm.tv/pic/cover/s/35/eb/104833_CAd6k.jpg",
            },
            "locked": false,
            "name": "Criminal Minds (Season 10)",
            "nameCN": "犯罪心理 第十季",
            "nsfw": false,
            "type": 6,
          },
          Object {
            "id": 148228,
            "images": Object {
              "common": "https://lain.bgm.tv/pic/cover/c/72/8c/148228_O5mS0.jpg",
              "grid": "https://lain.bgm.tv/pic/cover/g/72/8c/148228_O5mS0.jpg",
              "large": "https://lain.bgm.tv/pic/cover/l/72/8c/148228_O5mS0.jpg",
              "medium": "https://lain.bgm.tv/pic/cover/m/72/8c/148228_O5mS0.jpg",
              "small": "https://lain.bgm.tv/pic/cover/s/72/8c/148228_O5mS0.jpg",
            },
            "locked": false,
            "name": "Criminal Minds (Season 11)",
            "nameCN": "犯罪心理  第11季",
            "nsfw": false,
            "type": 6,
          },
          Object {
            "id": 184017,
            "locked": false,
            "name": "沙盒",
            "nameCN": "Bangumi Wiki 动画测试用沙盘",
            "nsfw": false,
            "type": 2,
          },
          Object {
            "id": 194118,
            "images": Object {
              "common": "https://lain.bgm.tv/pic/cover/c/5f/65/194118_5c9H0.jpg",
              "grid": "https://lain.bgm.tv/pic/cover/g/5f/65/194118_5c9H0.jpg",
              "large": "https://lain.bgm.tv/pic/cover/l/5f/65/194118_5c9H0.jpg",
              "medium": "https://lain.bgm.tv/pic/cover/m/5f/65/194118_5c9H0.jpg",
              "small": "https://lain.bgm.tv/pic/cover/s/5f/65/194118_5c9H0.jpg",
            },
            "locked": false,
            "name": "Criminal Minds Season 12",
            "nameCN": "犯罪心理 第12季",
            "nsfw": false,
            "type": 6,
          },
          Object {
            "id": 208783,
            "images": Object {
              "common": "https://lain.bgm.tv/pic/cover/c/c7/0d/208783_0J0TH.jpg",
              "grid": "https://lain.bgm.tv/pic/cover/g/c7/0d/208783_0J0TH.jpg",
              "large": "https://lain.bgm.tv/pic/cover/l/c7/0d/208783_0J0TH.jpg",
              "medium": "https://lain.bgm.tv/pic/cover/m/c7/0d/208783_0J0TH.jpg",
              "small": "https://lain.bgm.tv/pic/cover/s/c7/0d/208783_0J0TH.jpg",
            },
            "locked": false,
            "name": "Stranger Things Season 2",
            "nameCN": "怪奇物语 第二季",
            "nsfw": false,
            "type": 6,
          },
        ],
        "2": Array [],
      },
    },
    "6": Object {
      "counts": Object {
        "1": 0,
        "2": 0,
        "3": 0,
        "4": 0,
        "5": 0,
      },
      "details": Object {
        "1": Array [],
        "2": Array [],
      },
    },
  },
}
`;
